### PR TITLE
Add support for rootless network-aliases and static ip/mac

### LINKED ISSUE #7842: "rootless containers and pods cannot be assigned static IP addresses" (podman-run, rootless, CNI)
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

Assignment of a static IP address to rootless CNI  when running `podman run` fails.
If the failure is expected it should be documented in the documentation for the
`--ip` command-line option in the manual page for `podman run` 
https://github.com/containers/podman/blob/master/docs/source/markdown/podman-run.1.md

 

**Steps to reproduce the issue:**

1.

```
[erik@laptop ~]$ podman network create test
/home/erik/.config/cni/net.d/test.conflist
[erik@laptop ~]$ podman network inspect test
[
  {
    "cniVersion": "0.4.0",
    "name": "test",
    "plugins": [
      {
        "bridge": "cni-podman5",
        "hairpinMode": true,
        "ipMasq": true,
        "ipam": {
          "ranges": [
            [
              {
                "gateway": "10.88.7.1",
                "subnet": "10.88.7.0/24"
              }
            ]
          ],
          "routes": [
            {
              "dst": "0.0.0.0/0"
            }
          ],
          "type": "host-local"
        },
        "isGateway": true,
        "type": "bridge"
      },
      {
        "capabilities": {
          "portMappings": true
        },
        "type": "portmap"
      },
      {
        "backend": "",
        "type": "firewall"
      },
      {
        "domainName": "dns.podman",
        "type": "dnsname"
      }
    ]
  }
]
[erik@laptop ~]$ podman run --rm -ti --network test --ip=10.88.7.5 docker.io/library/alpine 
Error: invalid config provided: rootless containers and pods cannot be assigned static IP addresses
[erik@laptop ~]$ 
```

**Describe the results you received:**

**Describe the results you expected:**

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
podman version 2.1.1
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.16.1
  cgroupManager: cgroupfs
  cgroupVersion: v1
  conmon:
    package: conmon-2.0.21-1.el8.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.21, commit: 3460cd1ad859a79bd27df1714f39c76926ac1b39-dirty'
  cpus: 16
  distribution:
    distribution: '"centos"'
    version: "8"
  eventLogger: journald
  hostname: laptop.example.com (manualleditted to mask information)
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1008
      size: 1
    - container_id: 1
      host_id: 427680
      size: 65536
    uidmap:
    - container_id: 0
      host_id: 1008
      size: 1
    - container_id: 1
      host_id: 427680
      size: 65536
  kernel: 4.18.0-193.19.1.el8_2.x86_64
  linkmode: dynamic
  memFree: 63281815552
  memTotal: 67206131712
  ociRuntime:
    name: runc
    package: runc-1.0.0-145.rc91.git24a3cf8.el8.x86_64
    path: /usr/bin/runc
    version: 'runc version spec: 1.0.2-dev'
  os: linux
  remoteSocket:
    path: /run/user/1008/podman/podman.sock
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: slirp4netns-1.1.4-2.el8.x86_64
    version: |-
      slirp4netns version 1.1.4
      commit: b66ffa8e262507e37fca689822d23430f3357fe8
      libslirp: 4.3.1
      SLIRP_CONFIG_VERSION_MAX: 3
  swapFree: 33810280448
  swapTotal: 33810280448
  uptime: 41h 17m 7.7s (Approximately 1.71 days)
registries:
  search:
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org
  - docker.io
store:
  configFile: /home/erik/.config/containers/storage.conf
  containerStore:
    number: 44
    paused: 0
    running: 5
    stopped: 39
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-1.1.2-1.el8.x86_64
      Version: |-
        fusermount3 version: 3.2.1
        fuse-overlayfs: version 1.1.0
        FUSE library version 3.2.1
        using FUSE kernel interface version 7.26
  graphRoot: /home/erik/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: xfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 143
  runRoot: /run/user/1008/containers
  volumePath: /home/erik/.local/share/containers/storage/volumes
version:
  APIVersion: 2.0.0
  Built: 1601258926
  BuiltTime: Mon Sep 28 04:08:46 2020
  GitCommit: ""
  GoVersion: go1.13.15
  OsArch: linux/amd64
  Version: 2.1.1


```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-2.1.1-4.el8.x86_64
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**

### LINKED ISSUE #8567: --network-alias doesn't seem to work
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind bug

**Description**

When I try to use the new (since 2.2.0) network alias feature, it doesn't seem to work.  

**Steps to reproduce the issue:**

```bash
podman network create net1
podman run -d --rm --name alpine1 --network net1 --network-alias a1 alpine tail -f /dev/null
podman run --rm --network net1 alpine nslookup a1     
```

**Describe the results you received:**
```
Server:		10.88.2.1
Address:	10.88.2.1:53

** server can't find a1: NXDOMAIN

** server can't find a1: NXDOMAIN
```

**Describe the results you expected:**

I would like to be able to resolve `a1` to the IP of the first container.

**Additional information you deem important (e.g. issue happens only occasionally):**

**Output of `podman version`:**

```
Version:      2.2.0
API Version:  2.1.0
Go Version:   go1.15.5
Git Commit:   db1d2ff111ee9b012779ff3a5279a982520ccda4
Built:        Tue Dec  1 23:59:35 2020
OS/Arch:      linux/amd64
```

**Output of `podman info --debug`:**

```
host:
  arch: amd64
  buildahVersion: 1.18.0
  cgroupManager: systemd
  cgroupVersion: v2
  conmon:
    package: Unknown
    path: /usr/bin/conmon
    version: 'conmon version 2.0.21, commit: 35a2fa83022e56e18af7e6a865ba5d7165fa2a4a'
  cpus: 4
  distribution:
    distribution: arch
    version: unknown
  eventLogger: journald
  hostname: carbon
  idMappings:
    gidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65535
    uidmap:
    - container_id: 0
      host_id: 1000
      size: 1
    - container_id: 1
      host_id: 100000
      size: 65535
  kernel: 5.9.11-arch2-1
  linkmode: dynamic
  memFree: 4999450624
  memTotal: 8216764416
  ociRuntime:
    name: crun
    package: Unknown
    path: /usr/bin/crun
    version: |-
      crun version 0.16
      commit: eb0145e5ad4d8207e84a327248af76663d4e50dd
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  os: linux
  remoteSocket:
    path: /run/user/1000/podman/podman.sock
  rootless: true
  slirp4netns:
    executable: /usr/bin/slirp4netns
    package: Unknown
    version: |-
      slirp4netns version 1.1.7
      commit: e62caa08b78f3e662422bd7bfbcd2df3d12dcab1
      libslirp: 4.4.0
      SLIRP_CONFIG_VERSION_MAX: 3
      libseccomp: 2.5.1
  swapFree: 8589930496
  swapTotal: 8589930496
  uptime: 10m 35.75s
registries:
  search:
  - docker.io
  - registry.fedoraproject.org
  - quay.io
  - registry.access.redhat.com
  - registry.centos.org
store:
  configFile: /home/tino/.config/containers/storage.conf
  containerStore:
    number: 7
    paused: 0
    running: 2
    stopped: 5
  graphDriverName: overlay
  graphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: Unknown
      Version: |-
        fusermount3 version: 3.10.0
        fuse-overlayfs: version 1.3
        FUSE library version 3.10.0
        using FUSE kernel interface version 7.31
  graphRoot: /home/tino/.local/share/containers/storage
  graphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  imageStore:
    number: 95
  runRoot: /run/user/1000/containers
  volumePath: /home/tino/.local/share/containers/storage/volumes
version:
  APIVersion: 2.1.0
  Built: 1606863575
  BuiltTime: Tue Dec  1 23:59:35 2020
  GitCommit: db1d2ff111ee9b012779ff3a5279a982520ccda4
  GoVersion: go1.15.5
  OsArch: linux/amd64
  Version: 2.2.0
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
pacman -Q podman
podman 2.2.0-1
```

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

Yes

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Local laptop. The rootless-cni-infra image is `quay.io/libpod/rootless-cni-infra@sha256:304742d5d221211df4ec672807a5842ff11e3729c50bc424ea0cea858f69d7b7`

### PR DESCRIPTION
**Pull request by Luap99**
Make sure we pass the ip and mac address as CNI_ARGS to
the cnitool which is executed in the rootless-cni-infra container.

Fixes #7842

Make sure we pass the network aliases as capability args to the
cnitool in the rootless-cni-infra container. Also update the
dnsname plugin in the cni-infra container.

Fixes #8567

### CODE REVIEW

File: contrib/rootless-cni-infra/rootless-cni-infra [hunk 1]
```diff
@@ -46,6 +49,16 @@ cmd_entrypoint_alloc() {
...
 	CNI_ARGS="IgnoreUnknown=1;K8S_POD_NAME=${K8S_POD_NAME}"
+	if [ "$IP" ]; then
+		CNI_ARGS="$CNI_ARGS;IP=${IP}"
+	fi
+	if [ "$MAC" ]; then
+		CNI_ARGS="$CNI_ARGS;MAC=${MAC}"
+	fi
+	if [ "$CAP_ARGS" ]; then
+		CAP_ARGS="$CAP_ARGS"
+		export CAP_ARGS
```
Comment (AkihiroSuda) [inline contrib/rootless-cni-infra/rootless-cni-infra:60]: nit: export should be in L64: `export CNI_ARGS CNI_IFNAME`

**Comment by trusch** [comment]
@Luap99 Hello! I just wanted to ask if I can help anyhow on this PR? I'm really interested in this feature and would love to do my part to make it happen!

**Comment by tmds** [comment]
I'm looking forward to use this! @Luap99 are you still working on this?

**Comment by Luap99** [comment]
I focus on #8910. This PR also contains a fix for this.

**Comment by Luap99** [comment]
@baude @mheon @vrothberg @rhatdan This is something that users would like to have for 3.0. It should also help docker-compose support for rootless users. If you agree that we should get this into 3.0 I would rebase but I need somebody to build and push the new rootless-cni-infra image to quay.io.

#8910 would also fix this but it is not ready.

**Comment by mheon** [comment]
Agree - if you can rebase, we can merge + backport so this is fixed for 3.0

**Comment by Luap99** [comment]
OK. This one should be ready to build the image.

**Comment by mheon** [comment]
LGTM on my end. @containers/podman-maintainers  PTAL and merge, we can get the image built subsequently.
@Luap99 OK to remove the WIP?

File: libpod/rootless_cni_linux.go [hunk 2]
```diff
@@ -25,7 +25,7 @@ import (
 
 // Built from ../contrib/rootless-cni-infra.
 var rootlessCNIInfraImage = map[string]string{
-	"amd64": "quay.io/libpod/rootless-cni-infra@sha256:304742d5d221211df4ec672807a5842ff11e3729c50bc424ea0cea858f69d7b7", // 3-amd64
+	"amd64": "quay.io/luap99/rootless-cni-infra@sha256:fcc48fa68be51ae177d8c4c865c448d93531fa874362b710f2d64b117687268c", // 5-amd64
```
Comment (Luap99) [inline libpod/rootless_cni_linux.go:28]: @mheon I need the image before we merge. I have to update this line with the correct digest.

Comment (mheon) [reply libpod/rootless_cni_linux.go:28]: Ack, let me do that.

Comment (mheon) [reply libpod/rootless_cni_linux.go:28]: Built and pushed as `sha256:adf352454666f7ce9ca3e1098448b5ee18f89c4516471ec99447ec9ece917f36`

Comment (Luap99) [reply libpod/rootless_cni_linux.go]: Thanks

**Comment by rhatdan** [comment]
Great job @Luap99 
/lgtm

**Pull request merged**